api/vector_tiles.py

```diff
@@ -58,7 +58,7 @@ def index(self, zoom, x, y):
         mapdb = cherrypy.request.app.config['DB']['map']
         d = mapdb.tables.style.data
 
-        q = sa.select([d.c.rels, d.c.allshields.label('shields'),
+        q = sa.select([d.c.rels, d.c.allrels, d.c.allshields.label('shields'),
```

Comment:
If I understand correctly, you would not add another column to the defstyle table but compute the information in vector_tiles using an sql command?

With an arbitrary tile selected and a new index on hierarchy.child, the SQL plan costs:

```sql
EXPLAIN SELECT cycling.defstyle.rels, 
    (select array(select cycling.hierarchy.parent FROM cycling.hierarchy WHERE cycling.hierarchy.child=ANY(cycling.segments.rels))) as hierarchyrels,
    cycling.defstyle.allshields AS shields, 
    cycling.defstyle.network, 
    cycling.defstyle.style, 
    cycling.defstyle.class 
FROM 
    cycling.defstyle, 
    cycling.segments
WHERE cycling.defstyle.geom && (bbox)
and cycling.defstyle.id=cycling.segments.id 
ORDER BY cycling.defstyle.id;

 Sort  (cost=1033.39..1033.43 rows=17 width=176)
   Sort Key: defstyle.id
   ->  Nested Loop  (cost=4.69..1033.04 rows=17 width=176)
         ->  Bitmap Heap Scan on defstyle  (cost=4.41..68.68 rows=17 width=143)
               Recheck Cond: (geom && '0xxx'::geometry)
               ->  Bitmap Index Scan on idx_defstyle_geom  (cost=0.00..4.40 rows=17 width=0)
                     Index Cond: (geom && '0xxx'::geometry)
         ->  Index Scan using segments_pkey on segments  (cost=0.29..7.83 rows=1 width=41)
               Index Cond: (id = defstyle.id)
         SubPlan 2
           ->  Result  (cost=48.87..48.88 rows=1 width=0)
                 InitPlan 1 (returns $1)
                   ->  Index Scan using hierarchy_child_idx on hierarchy  (cost=0.29..48.87 rows=11 width=8)
                         Index Cond: (child = ANY (segments.rels))
```

The current implementation has a cost of
```sql
EXPLAIN SELECT cycling.defstyle.rels, cycling.defstyle.allshields AS shields, cycling.defstyle.network, cycling.defstyle.style, cycling.defstyle.class FROM cycling.defstyle                                                                                                                                
WHERE cycling.defstyle.geom && (bbox) ORDER BY cycling.defstyle.id;

 Sort  (cost=69.02..69.07 rows=17 width=115)
   Sort Key: id
   ->  Bitmap Heap Scan on defstyle  (cost=4.41..68.68 rows=17 width=115)
         Recheck Cond: (geom && '0xxx'::geometry)
         ->  Bitmap Index Scan on idx_defstyle_geom  (cost=0.00..4.40 rows=17 width=0)
               Index Cond: (geom && '0xxx'::geometry)
```

Is a cost of 1000 okay? How much difference makes a cost factor of 10 in runtime? Or: can I simplify the query?

Should the column contain all relations except top ones? In analogy to `defstyle.inrshields` vs `allshields` I've duplicated the relation id. But as you said, the top relation can be easily removed or concatenated...